smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -63,93 +92,228 @@ class DragonBackend:
+        print("-------------------------------------------------------------\n")
+
+    @property
+    def should_shutdown(self) -> bool:
+        return self._shutdown_requested
+
+    def _initialize_hosts(self) -> None:
+        with self._hostlist_lock:
+            self._hosts: t.List[str] = sorted(
+                Node(node).hostname for node in System().nodes
+            )
+            self._free_hosts: t.Deque[str] = collections.deque(self._hosts)
+            self._allocated_hosts: t.Dict[str, str] = {}
+
+    def _request_is_satisfiable(self, request: DragonRunRequest) -> bool:
+        """Check if request can be satisfied with resources
+        available in the allocation.
+        Currently only checks for total number of nodes,
+        in the future it will also look at other constraints
+        such as memory, accelerators, and so on.
+        """
+        return request.nodes <= len(self._hosts)
+
+    def _allocate_step(
+        self, step_id: str, request: DragonRunRequest
+    ) -> t.Optional[t.List[str]]:
+
+        num_hosts: int = request.nodes
+        with self._hostlist_lock:
+            if num_hosts <= 0 or num_hosts > len(self._free_hosts):
+                return None
+            to_allocate = []
+            for _ in range(num_hosts):
+                host = self._free_hosts.popleft()
+                self._allocated_hosts[host] = step_id
+                to_allocate.append(host)
+            return to_allocate
 
     def _get_new_id(self) -> str:
         with self._step_id_lock:
             self._step_id += 1
-            return str(self._step_id)
+            return str(uuid4())[-4:] + "-" + str(self._step_id)
 
     @functools.singledispatchmethod
     # Deliberately suppressing errors so that overloads have the same signature
     # pylint: disable-next=no-self-use
     def process_request(self, request: DragonRequest) -> DragonResponse:
-        raise TypeError("Unsure how to process a `{type(request)}` request")
+        raise TypeError(f"Unsure how to process a `{type(request)}` request")
 
     @process_request.register
     def _(self, request: DragonRunRequest) -> DragonRunResponse:
 
-        proc = TemplateProcess(
-            target=request.exe,
-            args=request.exe_args,
-            cwd=request.path,
-            env={**request.current_env, **request.env},
-            # stdout=Popen.PIPE,
-            # stderr=Popen.PIPE,
-        )
-
-        grp = ProcessGroup(restart=False, pmi_enabled=request.pmi_enabled)
-        grp.add_process(nproc=request.tasks, template=proc)
         step_id = self._get_new_id()
-        grp.init()
-        grp.start()
-        self._proc_groups[step_id] = (grp, grp.puids)
+        if not self._request_is_satisfiable(request):
+            message = f"Cannot satisfy request. Requested {request.nodes} nodes, "
+            message += f"but only {len(self._hosts)} nodes are available."
```

Comment:
@MattToast So, do you think that `_request_is_satisfiable(request)` should be returning the error message, and if it is empty we should assume it can be satisfied? Or should we return a `tuple(bool, str)` as in `(satisfiable, err_message`?

Reply:
Going with the second option for now. Also changed the name to the more English-sounding `_can_honor_request`.

Reply:
^^ Love it! yeah I'd recommend returning a tuple rather than having some sentinel value mean "good" and I like the name change!!